### ISSUE TITLE
Native batch_norm

### ISSUE THREAD

**Issue by t-vi** [pytorch/pytorch#9084]
I implemented batch_norm natively in ATen. It seems to do the right thing, but I think the existing tests are not comprehensive.

**Comment by zou3519** [comment]
What tests do you think are needed?

**Comment by t-vi** [comment]
My top two were that it runs (I did have an error in the reshaping of the bias that I only found in my own comparison with the existing code but that did not seem to trigger a test failure) and that grads work as expected.

**Comment by fmassa** [comment]
Wouldn't having a native implementation slow down significantly networks that have several batch norm layers?

**Comment by t-vi** [comment]
I'll have to benchmark it (maybe using resnet or so), but if you think it's an inherently bad idea, I'll just drop the idea.

**Comment by fmassa** [comment]
I think that it can potentially slow down quite a lot, specially given that we have cudnn and optimized c/cuda code for it

**Comment by t-vi** [comment]
No, I didn't want to throw out useful optimization, sorry if it sounded like that. I had understood that there was a general move from TH(CU)NN to ATen/native.
One thing I noticed is that the accumulation kernel in ATen/native/cpu doesn't differentiate between "data touching accumulation" and "further accumulation of intermediates", but that would be useful for an optimized variance. (I do believe that calculating mean and variance in a joint kernel would probably be a good idea for further optimization, too, but that is probably just me.)

**Comment by fmassa** [comment]
We are indeed moving more and more of our layers to ATen, and having batch norm entirely in ATen is definitely something that will happen at some point in time.
But there are still a few things missing in ATen that would make converting BN to ATen easy. The first thing that comes to my mind is a DeviceTensor, that allows us to index nd tensors in cuda kernels, and is used in BN on the GPU.
